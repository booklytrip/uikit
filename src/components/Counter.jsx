/**
 * @flow
 */

import React from 'react';
import classNames from 'classnames';
import { isNil } from 'lodash';

import { Button, Popup } from 'semantic-ui-react';
import popupWithCondition from '../lib/popupWithCondition';

type CounterSize = 'mini' | 'small' | 'large' | 'big' | 'huge' | 'massive';

type CounterProps = {
    style?: Object,
    inputStyle?: Object,
    placeholder?: string,
    value: string,
    formatValue?: Function,
    min?: number,
    max?: number,
    maxMessage?: string,
    size?: CounterSize,
    onChange?: Function,
};

const styles = {
    input: {
        textAlign: 'center',
        borderTopLeftRadius: 0,
        borderTopRightRadius: 0,
        borderBottomLeftRadius: 0,
        borderBottomRightRadius: 0,
        borderLeft: 0,
        borderRight: 0,
    },
    leftButton: {
        borderRadius: '0.28571429rem 0px 0px 0.28571429rem',
        borderWidth: 0,
    },
    rightButton: {
        borderRadius: '0px 0.28571429rem 0.28571429rem 0px',
        borderWidth: 0,
    },
    disabledButton: {
        opacity: 0.45,
    },
};

const ConditionalPopup = popupWithCondition(Popup);

/**
 * Return incremented value. If value is greater then max return max value.
 *
 * @param {Number} originalValue - The current value
 * @param {Number} max           - The max allowed value
 *
 * @return {Number}
 */
const incValue = (originalValue: number, max: number) => {
    if (max === undefined) {
        return originalValue + 1;
    }

    const value = originalValue || 0;
    return value + 1 < max ? value + 1 : max;
};

/**
 * Return decremented value. If value is less than min, return min value.
 *
 * @param {Number} originalValue - The current value
 * @param {Number} min           - The min allowed value
 *
 * @return {Number}
 */
const decValue = (originalValue: number, min: number) => {
    if (min === undefined) {
        return originalValue - 1;
    }

    const value = originalValue || min;
    return value - 1 > min ? value - 1 : min;
};

const Counter = ({
    style,
    inputStyle,
    value: originalValue,
    formatValue,
    placeholder,
    min,
    max,
    maxMessage,
    size,
    onChange,
}: CounterProps) => {
    const value = !isNil(originalValue) ? parseInt(originalValue, 10) : '';

    if (!formatValue) {
        formatValue = value => value;
    }

    return (
        <div className={classNames('ui action input', size)} style={style}>
            <Button
                style={Object.assign(
                    {},
                    styles.leftButton,
                    min === value && styles.disabledButton,
                )}
                onClick={e => {
                    e.preventDefault();
                    onChange(decValue(value, min));
                }}
                size={size}
                basic
            >
                -
            </Button>
            <input
                style={Object.assign({}, styles.input, inputStyle)}
                type="text"
                placeholder={placeholder}
                value={value}
                readOnly
            />
            <ConditionalPopup
                content={maxMessage}
                condition={value === max}
                position="top right"
                size="small"
                on="click"
                trigger={
                    <Button
                        style={Object.assign(
                            {},
                            styles.rightButton,
                            max === value && styles.disabledButton,
                        )}
                        onClick={e => {
                            e.preventDefault();
                            onChange(incValue(value || min, max));
                        }}
                        size={size}
                        basic
                    >
                        +
                    </Button>
                }
            />
        </div>
    );
};

export default Counter;
