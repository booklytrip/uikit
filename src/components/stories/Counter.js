import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import centered from '@kadira/react-storybook-decorator-centered';

import Counter from '../Counter.jsx';

storiesOf('Counter', module)
    .addDecorator(centered)
    .add('with placeholder', () => (
        <Counter placeholder="Enter value" onChange={action('onChange')} />
    ))
    .add('with value', () => (
        <Counter
            placeholder="Enter value"
            value={5}
            onChange={action('onChange')}
        />
    ))
    .add('message when max is reached', () => (
        <Counter
            placeholder="Enter value"
            value={1}
            max={1}
            maxMessage="The max value is reached"
            onChange={action('onChange')}
        />
    ))
    .add('different size', () => (
        <div>
            <p><Counter size="mini" placeholder="mini" /></p>
            <p><Counter size="small" placeholder="small" /></p>
            <p><Counter size="large" placeholder="large" /></p>
            <p><Counter size="big" placeholder="big" /></p>
            <p><Counter size="huge" placeholder="huge" /></p>
            <p><Counter size="massive" placeholder="massive" /></p>
        </div>
    ));
