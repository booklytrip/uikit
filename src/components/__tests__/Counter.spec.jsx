import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Counter from '../Counter';

jest.mock('../Popup.jsx');

describe('Counter', () => {
    it('renders without crashing', () => {
        shallow(<Counter />);
    });

    it('should call onChange event decreased value by 1', () => {
        const spy = jest.fn();
        const counter = shallow(<Counter value={3} onChange={spy} />);
        counter.find('Button').at(0).simulate('click');

        expect(spy).toHaveBeenCalledWith(2);
    });

    it('should call onChange event with increase value by 1', () => {
        const spy = jest.fn();
        const counter = shallow(<Counter value={3} onChange={spy} />);
        counter.find('Button').at(1).simulate('click');

        expect(spy).toHaveBeenCalledWith(4);
    });

    it('should decrese value to 0', () => {
        const spy = jest.fn();
        const counter = shallow(<Counter value={1} onChange={spy} />);
        counter.find('Button').at(0).simulate('click');

        expect(spy).toHaveBeenCalledWith(0);
    });

    it('should not decrease value lower then specified min value', () => {
        const spy = jest.fn();
        const counter = shallow(<Counter value={3} min={3} onChange={spy} />);
        counter.find('Button').at(0).simulate('click');

        expect(spy).toHaveBeenCalledWith(3);
    });

    it('should not increase value higher then specified max value', () => {
        const spy = jest.fn();
        const counter = shallow(<Counter value={3} max={3} onChange={spy} />);
        counter.find('Button').at(1).simulate('click');

        expect(spy).toHaveBeenCalledWith(3);
    });

    describe('Snapshots', () => {
        it('renders with specified value', () => {
            const rendered = renderer.create(<Counter value={5} />);
            expect(rendered.toJSON()).toMatchSnapshot();
        });

        it('renders with specified placeholder', () => {
            const rendered = renderer.create(
                <Counter placeholder="placeholder" />,
            );
            expect(rendered.toJSON()).toMatchSnapshot();
        });
    });
});
