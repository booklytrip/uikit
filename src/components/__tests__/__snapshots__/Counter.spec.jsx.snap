exports[`Counter Snapshots renders with specified placeholder 1`] = `
<div
  className="ui action input"
  style={undefined}>
  <button
    className="ui button basic"
    onClick={[Function]}
    style={
      Object {
        "borderRadius": "0.28571429rem 0px 0px 0.28571429rem",
        "borderWidth": 0,
      }
    }
    type={undefined}>
    -
  </button>
  <input
    placeholder="placeholder"
    readOnly={true}
    style={
      Object {
        "borderBottomLeftRadius": 0,
        "borderBottomRightRadius": 0,
        "borderLeft": 0,
        "borderRight": 0,
        "borderTopLeftRadius": 0,
        "borderTopRightRadius": 0,
        "textAlign": "center",
      }
    }
    type="text"
    value="" />
</div>
`;

exports[`Counter Snapshots renders with specified value 1`] = `
<div
  className="ui action input"
  style={undefined}>
  <button
    className="ui button basic"
    onClick={[Function]}
    style={
      Object {
        "borderRadius": "0.28571429rem 0px 0px 0.28571429rem",
        "borderWidth": 0,
      }
    }
    type={undefined}>
    -
  </button>
  <input
    placeholder={undefined}
    readOnly={true}
    style={
      Object {
        "borderBottomLeftRadius": 0,
        "borderBottomRightRadius": 0,
        "borderLeft": 0,
        "borderRight": 0,
        "borderTopLeftRadius": 0,
        "borderTopRightRadius": 0,
        "textAlign": "center",
      }
    }
    type="text"
    value={5} />
</div>
`;
